torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,307 @@
+          return false;
+        }
+      }
+    }
+    auto it = preservedAttrs_.find(attrModule._ivalue());
+    return it == preservedAttrs_.end() || !it->second.count(name);
+  }
+
+  void recordMutableAttrs(std::shared_ptr<Graph>& graph) {
+    std::stack<Block*> blocks({graph->block()});
+    std::unique_ptr<AliasDb> aliasDb =
+        torch::make_unique<AliasDb>(graph, /* isFrozen */ true);
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::SetAttr || n->kind() == prim::GetAttr) {
+          auto name = n->s(attr::name);
+          auto inputNode = n->inputs()[0]->node();
+          auto attrModule = module_;
+          if (!findConstantAttr(inputNode, name, attrModule)) {
+            continue;
+          }
+          if (n->kind() == prim::SetAttr || aliasDb->hasOutputWriters(n)) {
+            GRAPH_DEBUG(
+                n->kind() == prim::GetAttr ? "attribute: " + name + " in %" +
+                        n->outputs()[0]->debugName() + " has inplace writer"
+                                           : "");
+            preservedAttrs_[attrModule._ivalue()].insert(name);
+          }
+        }
+      }
+    }
+  }
+
+  std::set<std::string> recordReferencedAttrs(std::shared_ptr<Graph>& graph) {
+    std::stack<Block*> blocks({graph->block()});
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* subBlock : n->blocks()) {
+          blocks.push(subBlock);
+        }
+        if (n->kind() == prim::GetAttr) {
+          auto& name = n->s(attr::name);
+          if (module_.hasattr(name)) {
+            preservedAttrs_[module_._ivalue()].insert(name);
+          }
+        }
+      }
+    }
+    auto it = preservedAttrs_.find(module_._ivalue());
+    if (it != preservedAttrs_.end())
+      return it->second;
+    else
+      return std::set<std::string>();
+  }
+
+  // If Module is in eval mode, detach and override requires_grad tensor
+  // attributes to enable freezing.
+  bool shouldFoldAttr(IValue& attr, std::string& name, bool isEval) {
+    if (attr.isTensor()) {
+      auto t = attr.toTensor();
+      if (t.requires_grad()) {
+        t = autograd::as_variable_ref(t).detach();
+        attr = IValue(t);
+        if (isEval) {
+          t.set_requires_grad(false);
+        } else {
+          return false;
+        }
+      }
+    } else if (attr.isTensorList()) {
+      c10::List<at::Tensor> lst = std::move(attr).toTensorList();
+      for (size_t i = 0; i < lst.size(); ++i) {
+        auto t = autograd::as_variable_ref(lst.extract(i)).detach();
+        lst.set(i, t);
+        if (isEval) {
+          t.set_requires_grad(false);
+        } else {
+          return false;
+        }
+      }
+      attr = std::move(lst);
```

Comment:
we need to make sure this logic is sufficient, e.g. does it cover tuples that may contain tensors?

Reply:
Done